### PR TITLE
feat: Add support for ISO 32000-2 AES256 encryption

### PR DESCRIPTION
**Pull request by Darkheir**
**Pull request**

ISO 32000-2 is bringing a new password derivation algorithm that must be applied before using the key to decrypt the document.

This algorithm is needed for documents using the Version 5, Revision 6. PDF 2.0 is also deprecating Version 5 Revision 5 because of security issues.

This PR aims to support this new algorithm.

**How Has This Been Tested?**

Added tests

**Checklist**

- [x] I have added tests that prove my fix is effective or that my feature 
  works
- [x] I have added docstrings to newly created methods and classes
- [x] I have optimized the code at least one time after creating the initial 
  version
- [x] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [x] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [x] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

**Comment by pietermarsman** [comment]
Hi @Darkheir,

Thanks for this works. 

If I understand correctly this adds functionality that is added with the PDF 2.0 specification. 

To do a sensible review I want to look up the PDF 2.0 specification but it seems that you have [to pay for it](https://superuser.com/questions/1363832/is-iso-32000-22017-pdf-2-0-really-a-non-open-and-non-free-format). Do you know if that is true? And where did you get your copy of the 2.0 specification?

**Comment by Darkheir** [comment]
Hi @pietermarsman 

Sadly I didn't get access to the specification since you have to pay for it as you said.

I found a python project that implemented the feature: https://github.com/MatthiasValvekens/pyHanko/blob/master/pyhanko/pdf_utils/crypt.py

The code is based on this implementation, I only changed it to fit with the pdfminer code base.

I also added an encrypted pdf to test the implemented feature.

File: pdfminer/pdfdocument.py [hunk 1]
```diff
@@ -500,28 +500,74 @@ def get_cfm(self, name):
...
+        hash = self._password_hash(password, self.u_validation_salt)
+        if hash == self.u_hash:
+            hash = self._password_hash(password, self.u_key_salt)
+            cipher = Cipher(algorithms.AES(hash),
                             modes.CBC(b'\0' * 16),
                             backend=default_backend())
             return cipher.decryptor().update(self.ue)
         return None
 
+    def _password_hash(self, password, salt, vector=b""):
```
Comment (pietermarsman) [inline pdfminer/pdfdocument.py:519]: Perfectly named method! Very clear :) 

I prefer to use `None` as a value to signal that it should be ignored. I also prefer explicitly checking if `vector is None`. Can you change that?

**Review by pietermarsman — APPROVED** [review]
I'm going to assume that this works, because the tests work.

File: pdfminer/pdfdocument.py (hunk 1 above)
Comment (pietermarsman) [reply pdfminer/pdfdocument.py:519]: I wanted to fix this myself but I don't have permissions to push code to your branch

**Review by pietermarsman — CHANGES_REQUESTED** [review]
So this MR needs 3 more changes I think:
- [x] Use `None` instead of `b''`
- [ ] Fix CHANGELOG.md conflict
- [x] Include reference to original implementation as @MatthiasValvekens suggested

File: pdfminer/pdfdocument.py [hunk 2]
```diff
@@ -500,28 +500,74 @@ def get_cfm(self, name):
 
     def authenticate(self, password):
         password = password.encode('utf-8')[:127]
```
Comment (MatthiasValvekens) [inline pdfminer/pdfdocument.py:502]: The password needs to be passed through SASLprep for Unicode normalisation. If not, the authentication procedure could produce incorrect output.

File: pdfminer/pdfdocument.py [hunk 3]
```diff
@@ -500,28 +500,74 @@ def get_cfm(self, name):
...
+        hash = sha256(password)
+        hash.update(salt)
+        if vector:
+            hash.update(vector)
+        return hash.digest()
+
+    def _r6_password(self, password, salt, vector):
+        """
+        Compute the password for revision 6
+        """
```
Comment (MatthiasValvekens) [inline pdfminer/pdfdocument.py:540]: As I stated higher up in the thread, this is essentially a reformatted version of https://github.com/MatthiasValvekens/pyHanko/blob/b066f22bf35b60be9a8f3d8c0fe01d22850671bc/pyhanko/pdf_utils/crypt.py#L234-L263 (restating it here to make the reference more precise).

I'd appreciate a license note in accordance with the terms of the MIT License, thanks :)

File: pdfminer/pdfdocument.py (hunk 2 above)
Comment (MatthiasValvekens) [reply pdfminer/pdfdocument.py:502]: See [here](https://github.com/pdf-association/safedocs/tree/main/Unicode%20passwords) for relevant test docs.

**Review by MatthiasValvekens — CHANGES_REQUESTED** [review]
I also added a couple of comments.

Another general one: correct me if I'm wrong, but I don't think you're validating the integrity of the **Perms** entry anywhere in the authentication process. Was that intentional? It's part of the key retrieval/password validation procedure for revision 6 security handlers. Right now, it doesn't provide much in the way of integrity protection, but If nothing else, it's a good sanity check.

File: pdfminer/pdfdocument.py (hunk 3 above)
Comment (Darkheir) [reply pdfminer/pdfdocument.py:540]: I added the license and an acknowledgment section in the README

File: pdfminer/pdfdocument.py (hunk 1 above)
Comment (Darkheir) [reply pdfminer/pdfdocument.py]: Done :-)

**Review by pietermarsman — APPROVED** [review]

**Comment by pietermarsman** [comment]
@Darkheir @MatthiasValvekens Thanks for the work and support on this!

**Comment by MartinThoma** [comment]
@Darkheir would you be interested in adding modern encryption support to PyPDF2 as well?

https://github.com/py-pdf/PyPDF2

We recently added decryption support. Only the encryption is missing

**Pull request merged**